networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/methods/BeaconBlocksByRangeMessageHandler.java

```diff
@@ -0,0 +1,88 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.jvmlibp2p.rpc.methods;
+
+import static com.google.common.primitives.UnsignedLong.ONE;
+import static com.google.common.primitives.UnsignedLong.ZERO;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.networking.libp2p.rpc.BeaconBlocksByRangeRequestMessage;
+import tech.pegasys.artemis.networking.p2p.jvmlibp2p.Peer;
+import tech.pegasys.artemis.networking.p2p.jvmlibp2p.rpc.LocalMessageHandler;
+import tech.pegasys.artemis.networking.p2p.jvmlibp2p.rpc.ResponseCallback;
+import tech.pegasys.artemis.networking.p2p.jvmlibp2p.rpc.RpcException;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.Store;
+
+public class BeaconBlocksByRangeMessageHandler
+    implements LocalMessageHandler<BeaconBlocksByRangeRequestMessage, BeaconBlock> {
+  private static final org.apache.logging.log4j.Logger LOG = LogManager.getLogger();
+
+  private final ChainStorageClient storageClient;
+
+  public BeaconBlocksByRangeMessageHandler(final ChainStorageClient storageClient) {
+    this.storageClient = storageClient;
+  }
+
+  @Override
+  public void onIncomingMessage(
+      final Peer peer,
+      final BeaconBlocksByRangeRequestMessage message,
+      final ResponseCallback<BeaconBlock> callback) {
+    LOG.trace(
+        "Peer {} requested {} BeaconBlocks from chain {} starting at slot {} with step {}",
+        peer.getPeerId(),
+        message.getHeadBlockRoot(),
+        message.getStartSlot(),
+        message.getCount(),
+        message.getStep());
+    try {
+      sendMatchingBlocks(message, callback);
+      callback.completeSuccessfully();
+    } catch (final RpcException e) {
+      callback.completeWithError(e);
+    }
+  }
+
+  private void sendMatchingBlocks(
+      final BeaconBlocksByRangeRequestMessage message, final ResponseCallback<BeaconBlock> callback)
+      throws RpcException {
+    final Store store = storageClient.getStore();
+    if (store == null) {
+      return;
+    }
+    if (!storageClient.isIncludedInBestState(message.getHeadBlockRoot())) {
+      return;
+    }
+    final UnsignedLong bestSlot = storageClient.getBestSlot();
+    if (bestSlot == null) {
+      LOG.error("No best slot present despite having at least one canonical block");
+      throw RpcException.SERVER_ERROR;
+    }
+
+    UnsignedLong remainingBlocks = message.getCount();
+    for (UnsignedLong slot = message.getStartSlot();
+        !remainingBlocks.equals(ZERO) && slot.compareTo(bestSlot) <= 0;
+        slot = slot.plus(message.getStep())) {
```

Comment:
I think we need to check somewhere that `step > 0`, so we don't get stuck sending the same block over and over.

Reply:
ooo, good spot.  Done.